File: seminar-5/deque/python/deque.py
class Node:
    """Узел двусвязного списка для Deque."""

    def __init__(self, data, prev=None, next=None):
        """
        Создаёт новый узел.

        :param data: значение, хранящееся в узле
        :param prev: ссылка на предыдущий узел
        :param next: ссылка на следующий узел
        """
        self.data = data
        self.prev = prev
        self.next = next


class Deque:
    """Двусторонняя очередь (Deque) на основе двусвязного списка."""

    def __init__(self):
        """
        Создаёт пустой deque.
        """
        self.head = None  # левый конец
        self.tail = None  # правый конец
        self._size = 0

    def push_front(self, value):
        """
        Добавляет элемент в начало deque.

        :param value: значение для добавления
        """
        node = Node(value, prev=None, next=self.head)

        if self.head is None:
            self.head = self.tail = node
        else:
            self.head.prev = node
            self.head = node

        self._size += 1

    def push_back(self, value):
        """
        Добавляет элемент в конец deque.

        :param value: значение для добавления
        """
        node = Node(value, prev=self.tail, next=None)

        if self.tail is None:
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node

        self._size += 1

    def pop_front(self):
        """
        Удаляет и возвращает элемент из начала deque.

        :return: значение из начала deque
        :raises IndexError: если deque пуст
        """
        if self.is_empty():
            raise IndexError("pop_front from empty deque")

        value = self.head.data
        self.head = self.head.next

        if self.head is None:
            self.tail = None
        else:
            self.head.prev = None

        self._size -= 1
        return value

    def pop_back(self):
        """
        Удаляет и возвращает элемент из конца deque.

        :return: значение из конца deque
        :raises IndexError: если deque пуст
        """
        if self.is_empty():
            raise IndexError("pop_back from empty deque")

        value = self.tail.data
        self.tail = self.tail.prev

        if self.tail is None:
            self.head = None
        else:
            self.tail.next = None

        self._size -= 1
        return value

    def peek_front(self):
        """
        Возвращает элемент из начала deque без удаления.

        :return: значение из начала deque
        :raises IndexError: если deque пуст
        """
        if self.is_empty():
            raise IndexError("peek_front from empty deque")
        return self.head.data

    def peek_back(self):
        """
        Возвращает элемент из конца deque без удаления.

        :return: значение из конца deque
        :raises IndexError: если deque пуст
        """
        if self.is_empty():
            raise IndexError("peek_back from empty deque")
        return self.tail.data

    def is_empty(self):
        """
        Проверяет, пуст ли deque.
        """
        return self._size == 0

    def size(self):
        """
        Возвращает количество элементов в deque.
        """
        return self._size

    def __len__(self):
        return self._size

    def __bool__(self):
        return self._size != 0

    def __repr__(self):
        left = None if self.is_empty() else self.head.data
        right = None if self.is_empty() else self.tail.data
        return f"Deque(size={self._size}, front={left!r}, back={right!r})"

    def __iter__(self):
        """
        Итерация слева направо (от head к tail).
        """
        cur = self.head
        while cur is not None:
            yield cur.data
            cur = cur.next

    def to_list(self):
        """
        Преобразует deque в список слева направо.
        """
        return list(self)

File: seminar-5/deque/java/Deque.java
public class Deque<T> {

    private static class Node<T> {
        T data;
        Node<T> prev;
        Node<T> next;

        Node(T data, Node<T> prev, Node<T> next) {
            this.data = data;
            this.prev = prev;
            this.next = next;
        }
    }

    private Node<T> head; // левый конец
    private Node<T> tail; // правый конец
    private int size;

    public Deque() {
        head = null;
        tail = null;
        size = 0;
    }

    public void pushFront(T value) {
        Node<T> node = new Node<>(value, null, head);

        if (head == null) {
            head = tail = node;
        } else {
            head.prev = node;
            head = node;
        }
        size++;
    }

    public void pushBack(T value) {
        Node<T> node = new Node<>(value, tail, null);

        if (tail == null) {
            head = tail = node;
        } else {
            tail.next = node;
            tail = node;
        }
        size++;
    }

    public T popFront() {
        if (isEmpty()) {
            throw new IllegalStateException("pop_front from empty deque");
        }

        T value = head.data;
        head = head.next;

        if (head == null) {
            tail = null;
        } else {
            head.prev = null;
        }

        size--;
        return value;
    }

    public T popBack() {
        if (isEmpty()) {
            throw new IllegalStateException("pop_back from empty deque");
        }

        T value = tail.data;
        tail = tail.prev;

        if (tail == null) {
            head = null;
        } else {
            tail.next = null;
        }

        size--;
        return value;
    }

    public T peekFront() {
        if (isEmpty()) {
            throw new IllegalStateException("peek_front from empty deque");
        }
        return head.data;
    }

    public T peekBack() {
        if (isEmpty()) {
            throw new IllegalStateException("peek_back from empty deque");
        }
        return tail.data;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    @Override
    public String toString() {
        Object left = isEmpty() ? null : head.data;
        Object right = isEmpty() ? null : tail.data;
        return "Deque(size=" + size + ", front=" + left + ", back=" + right + ")";
    }
}
